extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/WriteMongoP.java

```diff
@@ -0,0 +1,627 @@
+                }
+                if (transactionUtility.usesTransactionLifecycle()) {
+                    checkNotNull(mongoTransaction, "there is no active transaction");
+                    mongoTransaction.addWrites(collectionKey, writes);
+                } else {
+                    collection.bulkWrite(writes);
+                }
+            }
+
+            for (int i = 0; i < items.size(); i++) {
+                inbox.remove();
+            }
+        } catch (Exception e) {
+            logger.info("Unable to process Mongo Sink: " + e.getMessage());
+            // not removing from inbox, so it will be retried
+        }
+    }
+
+    private Set<Object> queryForExistingIds(ArrayList<I> items, MongoCollection<?> collection) {
+        List<Object> identityValues = items.stream()
+                                           .map(documentIdentityFn)
+                                           .filter(Objects::nonNull)
+                                           .collect(toList());
+        Iterable<Document> foundIds = collection.aggregate(
+                asList(project(include("_id")), match(in("_id", identityValues))),
+                Document.class);
+        Set<Object> docsFound = new HashSet<>();
+        for (Document document : foundIds) {
+            docsFound.add(document.get("_id"));
+        }
+        return docsFound;
+    }
+
+    @Override
+    public void close() {
+        if (mongoClient != null) {
+            mongoClient.close();
+        }
+        transactionUtility.close();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean snapshotCommitPrepare() {
+        return transactionUtility.snapshotCommitPrepare();
+    }
+
+    @Override
+    public boolean snapshotCommitFinish(boolean success) {
+        return transactionUtility.snapshotCommitFinish(success);
+    }
+
+    @Override
+    public boolean complete() {
+        transactionUtility.afterCompleted();
+        return true;
+    }
+
+    @Override
+    protected void restoreFromSnapshot(@Nonnull Object key, @Nonnull Object value) {
+        transactionUtility.restoreFromSnapshot(key, value);
+    }
+
+    boolean reconnectIfNecessary() {
+        if (!isConnectionUp()) {
+            if (connectionRetryTracker.shouldTryAgain()) {
+                try {
+                    mongoClient = connectionSupplier.get();
+                    connectionRetryTracker.reset();
+                    return true;
+                } catch (Exception e) {
+                    logger.warning("Could not connect to MongoDB", e);
+                    connectionRetryTracker.attemptFailed();
+                    return false;
+                }
+            } else {
+                throw new JetException("cannot connect to MongoDB");
+            }
+        } else {
+            return true;
+        }
+    }
+
+    private boolean isConnectionUp() {
+        return mongoClient != null;
```

Comment:
Is it safe assumpotion? shouldn't we call some operation from mongoClient, maybe `getClusterDescription`?

Reply:
I've added listing of databases and getting the first element - that should indeed go over network